internal/controllers/vgmanager/wipe_devices.go

```diff
@@ -4,62 +4,96 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"time"
 
 	lvmv1alpha1 "github.com/openshift/lvm-operator/v4/api/v1alpha1"
+	"github.com/openshift/lvm-operator/v4/internal/controllers/constants"
 	"github.com/openshift/lvm-operator/v4/internal/controllers/vgmanager/dmsetup"
 	"github.com/openshift/lvm-operator/v4/internal/controllers/vgmanager/lsblk"
-	"github.com/openshift/lvm-operator/v4/internal/controllers/vgmanager/lvm"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-func (r *Reconciler) wipeDevicesIfNecessary(ctx context.Context, volumeGroup *lvmv1alpha1.LVMVolumeGroup, blockDevices []lsblk.BlockDevice, vgs []lvm.VolumeGroup) (bool, error) {
+func (r *Reconciler) wipeDevices(
+	ctx context.Context,
+	volumeGroup *lvmv1alpha1.LVMVolumeGroup,
+	blockDevices []lsblk.BlockDevice,
+) (bool, error) {
 	logger := log.FromContext(ctx)
 
-	if volumeGroup.Spec.DeviceSelector == nil || volumeGroup.Spec.DeviceSelector.ForceWipeDevicesAndDestroyAllData == nil || !*volumeGroup.Spec.DeviceSelector.ForceWipeDevicesAndDestroyAllData {
+	if !r.shouldWipeDevicesOnVolumeGroup(volumeGroup) {
+		logger.V(1).Info("skipping wiping devices as the volume group does not require it")
 		return false, nil
 	}
 
-	wiped := false
+	updated := false
+
 	for _, path := range volumeGroup.Spec.DeviceSelector.Paths {
-		diskName, err := evalSymlinks(path)
-		if err != nil {
-			return false, fmt.Errorf("unable to find symlink for disk path %s: %v", path, err)
-		}
-		if _, partOfVG := isDeviceAlreadyPartOfVG(vgs, diskName); partOfVG {
-			continue
-		}
-		deviceWiped, err := r.wipeDevice(ctx, diskName, blockDevices)
-		if err != nil {
+		if deviceWiped, err := r.wipeDevice(ctx, path, blockDevices); err != nil {
 			return false, fmt.Errorf("failed to wipe device %s: %w", path, err)
-		}
-		if deviceWiped {
-			wiped = true
+		} else if deviceWiped {
+			updated = true
 		}
 	}
 	for _, path := range volumeGroup.Spec.DeviceSelector.OptionalPaths {
-		diskName, err := evalSymlinks(path)
-		if err != nil {
-			logger.Info(fmt.Sprintf("skipping wiping optional device %s as unable to find symlink: %v", path, err))
-			continue
-		}
-		if _, partOfVG := isDeviceAlreadyPartOfVG(vgs, diskName); partOfVG {
-			continue
-		}
-		deviceWiped, err := r.wipeDevice(ctx, diskName, blockDevices)
-		if err != nil {
+		if deviceWiped, err := r.wipeDevice(ctx, path, blockDevices); err != nil {
 			logger.Info(fmt.Sprintf("skipping wiping optional device %s: %v", path, err))
+		} else if deviceWiped {
+			updated = true
 		}
-		if deviceWiped {
-			wiped = true
+	}
+
+	if updated {
+		if volumeGroup.Annotations[constants.DevicesWipedAnnotation] == "" {
+			volumeGroup.Annotations = make(map[string]string)
```

Comment:
Are we overriding all the annotations here?

Reply:
I had a brain lag and didnt compare for map==nil but for the key, that should be fixed now